File: libs/go/oe/natsx/envelope.go
// Package natsx publishes and consumes JetStream messages wrapped in the C1
// event envelope: subject "oe.<type>.<tenant_id>", header Nats-Msg-Id = id
// for dedupe, and durable pull consumers with retry-with-backoff and
// dead-lettering after MaxDeliver.
package natsx

import (
	"context"
	"encoding/hex"
	"encoding/json"
	"strings"
	"time"

	"github.com/google/uuid"
	"go.opentelemetry.io/otel/trace"
)

// occurredAtLayout renders occurred_at with exactly millisecond precision,
// as required by contracts/events/envelope.schema.json.
const occurredAtLayout = "2006-01-02T15:04:05.000Z07:00"

// Envelope is the C1 envelope shared by every JetStream message.
type Envelope struct {
	ID            string          `json:"id"`
	Type          string          `json:"type"`
	SchemaVersion int             `json:"schema_version"`
	TenantID      uuid.UUID       `json:"tenant_id"`
	OccurredAt    string          `json:"occurred_at"`
	Source        string          `json:"source"`
	ContactID     *uuid.UUID      `json:"contact_id"`
	TraceParent   string          `json:"trace_parent,omitempty"`
	Data          json.RawMessage `json:"data"`
}

// buildEnvelope assembles a new envelope for an outgoing event: UUID v7 id,
// occurred_at now (UTC, ms precision), source and trace_parent from ctx.
func buildEnvelope(ctx context.Context, source, eventType string, tenantID uuid.UUID, contactID *uuid.UUID, data any) (Envelope, error) {
	id, err := uuid.NewV7()
	if err != nil {
		return Envelope{}, err
	}
	payload, err := json.Marshal(data)
	if err != nil {
		return Envelope{}, err
	}
	return Envelope{
		ID:            id.String(),
		Type:          eventType,
		SchemaVersion: 1,
		TenantID:      tenantID,
		OccurredAt:    time.Now().UTC().Format(occurredAtLayout),
		Source:        source,
		ContactID:     contactID,
		TraceParent:   traceParent(ctx),
		Data:          payload,
	}, nil
}

// traceParent renders the W3C traceparent header (version 00) for the span
// active in ctx, or "" if ctx carries no valid span context.
func traceParent(ctx context.Context) string {
	sc := trace.SpanContextFromContext(ctx)
	if !sc.IsValid() {
		return ""
	}
	flags := "00"
	if sc.IsSampled() {
		flags = "01"
	}
	return "00-" + sc.TraceID().String() + "-" + sc.SpanID().String() + "-" + flags
}

// parseTraceParent parses a W3C traceparent header (version 00) into a
// remote trace.SpanContext. It returns ok=false for "" or a malformed value.
func parseTraceParent(tp string) (sc trace.SpanContext, ok bool) {
	parts := strings.Split(tp, "-")
	if len(parts) != 4 || parts[0] != "00" {
		return trace.SpanContext{}, false
	}
	traceID, err := trace.TraceIDFromHex(parts[1])
	if err != nil {
		return trace.SpanContext{}, false
	}
	spanID, err := trace.SpanIDFromHex(parts[2])
	if err != nil {
		return trace.SpanContext{}, false
	}
	flags, err := hex.DecodeString(parts[3])
	if err != nil || len(flags) != 1 {
		return trace.SpanContext{}, false
	}
	return trace.NewSpanContext(trace.SpanContextConfig{
		TraceID:    traceID,
		SpanID:     spanID,
		TraceFlags: trace.TraceFlags(flags[0]),
		Remote:     true,
	}), true
}

File: packages/ts-common/src/nats/envelope.ts
// C1 envelope (contracts/events/envelope.schema.json): published on subject
// `oe.<type>.<tenant_id>` with header `Nats-Msg-Id = id` for JetStream dedup.
import { randomBytes } from 'node:crypto';
import { context as otelContext, trace } from '@opentelemetry/api';

export interface EventEnvelope<T = unknown> {
  /** UUID v7, lowercase. Also the JetStream dedup key (`Nats-Msg-Id`). */
  id: string;
  /** Dotted event type, e.g. `contacts.upserted`. First segment picks the stream. */
  type: string;
  schema_version: number;
  tenant_id: string;
  /** RFC 3339 with exactly millisecond precision. */
  occurred_at: string;
  /** Name of the emitting service, e.g. `core`. */
  source: string;
  contact_id: string | null;
  /** W3C traceparent (version 00), when a span is active. */
  trace_parent?: string;
  data: T;
}

/** UUID v7 (RFC 9562): 48-bit ms timestamp, version 7, random tail. */
export function uuidv7(): string {
  const ms = BigInt(Date.now());
  const rnd = randomBytes(10);
  const b = Buffer.alloc(16);
  b[0] = Number((ms >> 40n) & 0xffn);
  b[1] = Number((ms >> 32n) & 0xffn);
  b[2] = Number((ms >> 24n) & 0xffn);
  b[3] = Number((ms >> 16n) & 0xffn);
  b[4] = Number((ms >> 8n) & 0xffn);
  b[5] = Number(ms & 0xffn);
  b[6] = 0x70 | (rnd[0]! & 0x0f);
  b[7] = rnd[1]!;
  b[8] = 0x80 | (rnd[2]! & 0x3f);
  rnd.copy(b, 9, 3, 10);
  const hex = b.toString('hex');
  return `${hex.slice(0, 8)}-${hex.slice(8, 12)}-${hex.slice(12, 16)}-${hex.slice(16, 20)}-${hex.slice(20)}`;
}

/** Current span's W3C `traceparent`, or `undefined` outside a span. */
export function currentTraceParent(): string | undefined {
  const span = trace.getSpanContext(otelContext.active());
  if (!span) return undefined;
  const flags = span.traceFlags.toString(16).padStart(2, '0');
  return `00-${span.traceId}-${span.spanId}-${flags}`;
}

/** Builds a C1 envelope, filling `id`, `occurred_at` and `trace_parent`. */
export function buildEnvelope<T>(
  source: string,
  type: string,
  tenantId: string,
  contactId: string | null,
  data: T,
  schemaVersion = 1
): EventEnvelope<T> {
  const envelope: EventEnvelope<T> = {
    id: uuidv7(),
    type,
    schema_version: schemaVersion,
    tenant_id: tenantId,
    occurred_at: new Date().toISOString(),
    source,
    contact_id: contactId,
    data,
  };
  const traceParent = currentTraceParent();
  if (traceParent) envelope.trace_parent = traceParent;
  return envelope;
}

/** Subject for an envelope of `type` under `tenantId` (C1). */
export function subjectFor(type: string, tenantId: string): string {
  return `oe.${type}.${tenantId}`;
}
